components/brave_news/browser/initialization_promise.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_news/browser/initialization_promise.h"
+
+#include <algorithm>
+#include <iterator>
+#include <string>
+#include <utility>
+
+#include "base/functional/bind.h"
+#include "base/functional/callback_forward.h"
+#include "base/location.h"
+#include "base/memory/raw_ref.h"
+#include "base/task/sequenced_task_runner.h"
+#include "base/time/time.h"
+#include "brave/components/brave_news/browser/channels_controller.h"
+#include "brave/components/brave_news/browser/publishers_controller.h"
+
+namespace brave_news {
+
+constexpr int kBackoffs[] = {1, 5, 10};
+constexpr size_t kNumBackoffs = std::size(kBackoffs);
+
+InitializationPromise::InitializationPromise(
+    size_t max_retries,
+    BraveNewsPrefManager& pref_manager,
+    PublishersController& publishers_controller)
+    : pref_manager_(pref_manager),
+      publishers_controller_(publishers_controller),
+      max_retries_(max_retries) {}
+
+InitializationPromise::~InitializationPromise() = default;
+
+void InitializationPromise::OnceInitialized(base::OnceClosure on_initialized) {
+  // This should only be called when Brave News is enabled.
+  CHECK(pref_manager_->IsEnabled());
+
+  if (state_ == State::kInitialized || state_ == State::kFailed) {
+    std::move(on_initialized).Run();
+    return;
+  }
+
+  on_initializing_prefs_complete_.Post(FROM_HERE, std::move(on_initialized));
+
+  // Only start initializing once.
+  if (state_ == State::kNone) {
+    Initialize();
+  }
+}
+
+void InitializationPromise::Initialize() {
+  state_ = State::kInitializing;
+
+  auto subscriptions = pref_manager_->GetSubscriptions();
+
+  // If things are already initialized, we're done!
+  if (!subscriptions.channels().empty() ||
+      !subscriptions.enabled_publishers().empty() ||
+      !subscriptions.disabled_publishers().empty() ||
+      !subscriptions.direct_feeds().empty()) {
+    state_ = State::kInitialized;
+    on_initializing_prefs_complete_.Signal();
+    return;
+  }
+
+  publishers_controller_->GetLocale(
+      subscriptions, base::BindOnce(&InitializationPromise::OnGotLocale,
+                                    weak_ptr_factory_.GetWeakPtr()));
+}
+
+void InitializationPromise::OnGotLocale(const std::string& locale) {
+  // Keep track of which attempt this is.
+  attempts_++;
```

Comment:
totally optional: ditto

```suggestion
  CHECK_EQ(State::kInitializing, state_);

  // Keep track of which attempt this is.
  attempts_++;
```